### PR TITLE
fix(menu): fix for sub menu positioning

### LINKED ISSUE #10260: [Bug]: Menu - sub-menus are not rendering correctly when there's not enough space below
### Package

carbon-components-react

### Browser

Chrome

### Package version

7.49.1

### React version

16.9.0

### Description

Sub menus (`MenuItem` inside `MenuItem`) is not rendering in correct place when there's not enough space below. Typically, sub menu is rendering below the parent item, but when there's no place for it, it tries to render upwards. Sadly, something's wrong with the position calculation and the menu is rendering a little to high:
image

This causes the problem with accessibility and makes the sub-menu unusable. Well, user can still use keyboard navigation, but if cursor will leave the menu sub-menu will disappear, so it's impossible to click on sub-menu items. 

GIF example:
demo

How it should work:
image

### CodeSandbox example

https://codesandbox.io/s/ecstatic-maxwell-l8pxt

### Steps to reproduce

Create a menu that has sub-menus. The sub-menu must be tall enough to render upwards (so, there's not enough space below). Please, see the Codepen example and and GIF that visualize the issue. 

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [X] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

### PR DESCRIPTION
**Pull request by jakubfaliszewski**
Closes #10260

Adds additional calculations for sub-menus, that are rendered upwards. It's based on menu spacing/sizes, so it's flexible and works correctly for different sizes (sm, md, lg, custom)

standard
<details>
  <summary>Complex menu demo</summary>
  
complex

</details>
<details>
  <summary>Large menu demo</summary>
  
lg

</details>

#### Changelog

**Fixed**

- positioning of sub-menus

#### Testing / Reviewing

Follow the step from GIFs and experiment a little with the styling.

### CODE REVIEW

**Comment by jakubfaliszewski** [comment]
> **DCO Assistant Lite bot:** Thanks for your submission! We ask that you sign our [Developer Certificate of Origin](https://github.com/carbon-design-system/carbon-dco/blob/main/dco.md) before we can accept your contribution. You can sign the DCO by adding a comment below using this text:
> 
> _**I have read the DCO document and I hereby sign the DCO.**_
> 
> You can retrigger this bot by commenting **recheck** in this Pull Request

I have read the DCO document and I hereby sign the DCO.

**Comment by jakubfaliszewski** [comment]
recheck

**Comment by jakubfaliszewski** [comment]
I have read the DCO document and I hereby sign the DCO.

**Review by joshblack — APPROVED** [review]

**Comment by joshblack** [comment]
cc @janhassel

File: packages/react/src/components/Menu/_utils.js [hunk 1]
```diff
@@ -135,7 +147,9 @@ export function getPosition(
   elementDimensions,
   targetBoundaries,
   containerBoundaries,
-  preferredDirection = 1
+  preferredDirection = 1,
+  isRootLevel,
+  element
```
Comment (janhassel) [inline packages/react/src/components/Menu/_utils.js:154]: Looks great, thanks for the bug fix! Would you mind adding the two new parameters to the JSDoc annotation above?

**Review by janhassel — CHANGES_REQUESTED** [review]

File: packages/react/src/components/Menu/_utils.js (hunk 1 above)
Comment (jakubfaliszewski) [reply packages/react/src/components/Menu/_utils.js:154]: Forgot bout that, thanks 👍

**Review by dakahn — APPROVED** [review]

**Comment by jakubfaliszewski** [comment]
@all-contributors please add @jakubfaliszewski  for code

**Pull request merged**